### ISSUE TITLE
Global ctags file

### ISSUE THREAD

**Issue by vinsworldcom** [gstavi/npp#1]
Love the plugin, I've actually forked and made some "improvements".  One I'm contemplating is adding an option for a "global" ctags database file that would be searched if the current `tags` file does not find a match.  For example:

```
myproject
  main.cpp
  main.h
  tags
```

If I lookup `printf` it fails to find it since I don't define `printf` in my files, but I use `#include <stdio.h>` and use `printf`.  So I'd like to in the settings file perhaps add a 'GlobalCtagsFile' that I would create by `ctags -R` on my c/include directory for example.  Then, if I lookup `printf` when in Notepad++, the local directory `tags` file would come up empty, but before the search window is displayed, the none-found result would trigger a second tags search in the configured "global" 'tags' file.  If no match there, then the empty window would display.

The idea could be expanded for a few "global" ctags files, only 1 being "active" at a given time (I'm thinking like a dropdown combo box in the TagLEET window so default is 'None' for normal backward compatible operations, but then the config file would host maybe 3 to 5 options for global files - anyone of which could be selected as the global backup.  You may create 1 for C++, 1 for Perl, 1 for Python, etc. and just select that global one for the current project you're working.

What do you think?

Cheers.

**Comment by gstavi** [comment]
I had quite a lot of ideas for "Next Generation TagLEET" but I am too lazy to start with them. Note to self, better document them
Maybe I need to somehow turn it into a true open source project that allow external contributions.

Regarding your idea:
First step is obviously adding a persistent configuration to TagLEET.

Similar requests was brought before. A flexible way I think of is by adding virtual mappings of **dir --> tags**.
Example:
```
c:\some\path\myproject  -->  c:\users\myself\tags-repo\tags
c:\some\path            -->  c:\users\myself\tags-repo\more.tags
c:\                     -->  c:\gloabl\tags
c:\                     -->  c:\users\myself\tags-repo\tags\custom.tags
c:\                     -->  c:\users\myself\tags-repo\tags\manual.tags
```
So basically when TagLEET scan directories backward from the files full path it first checks for physical tags file at the location and than for virtual mapping.
That way a user can map global code location for himself without having to put tags files at the global locations. The virtual mappings will use global location as index but will reference tags files at the users private locations.

If we go that way then we can add extra settings of whether all tags files are always searched (probably good for default) or does it stop after first match, or something else.

**Comment by vinsworldcom** [comment]
My C++ skills are awful so my hacking at your plugin has resulting in some "improvements" (as I see them) albeit with some pretty horrendous code :-)  I was thinking more "small-time" than bigger picture that you're suggesting.  What I tried was to simply add a hard-coded global tags file location (figured if it worked, I could make it user-settable) to the TagLookupContext class, next to:

tag_leet_app.h:
```
class TagLookupContext
{
public:
  TagLookupContext(NppCallContext *in_NppC, const char *in_TagsFilePath, const char *in_GlobalTagsFilePath);
  ~TagLookupContext() {};
  void GetLineNumFromTag(bool PrefixMatch, TagList *tl) const;

  NppCallContext *NppC;
  const char *TagsFilePath;
  const char *GlobalTagsFilePath;
[...]
```

Then hardcode that in:

tag_leet_app.cpp:
```
void TagLeetApp::LookupTag()
{
  TL_ERR err;
  TlAppSync Sync(this);
  NppCallContext NppC(this);
  char TagsFilePath[TL_MAX_PATH];
  char GlobalTagsFilePath[TL_MAX_PATH] = "C:\\usr\\share\\ctags\\tags.mingw";
  TCHAR Msg[2048];
  int i;
[...]
  TagLookupContext TLCtx(&NppC, TagsFilePath, GlobalTagsFilePath);
[...]
```

Then in `PopulateTagList()`, instead of returning `err` on return from failed `PopulateTagListHelper()`, I have a `PopulateTagListHelperGlobal()` called which uses the global tags file.

I made all the other necessary declarations and changes to get this to compile fine.  And testing with some debugging message boxes or outfile logging shows that the first search in the "local" file fails and then the subsequent search in the global file seems to recognize the global file, but turns up nothing as well - even though I'm purposefully searching for a tag not in the local file that does exist in the global one.

I'm sure I'm missing something about maybe caching tags or tag file locations or something.

Cheers.

**Comment by vinsworldcom** [comment]
Found it!  It was:

.\tag_engine\tag_list.cpp:
```
92:	  if (cache == NULL)
93:	  {
94:	    err = tf.Init(TagsFilePath);
95:	    if (err)
96:	      return err;
97:	    cache = &tf;
98:	  }
```

Just comment out the `if` statement.  The `cache` must have been saving something between my subsequent calls to `TList.Create()` in both my `PopulateTagListHelper()` and `PopulateTagListHelperGlobal()`.

I have a simple Settings dialogue with all my added options including a text box to type a filename for a global tags file to be used.  Nothing in the text box means just operate as usual with the local tags file.

Cheers.
